lnbits/decorators.py

```diff
@@ -220,8 +243,20 @@ async def require_invoice_key(
         return wallet
 
 
-async def check_user_exists(usr: UUID4) -> User:
-    g().user = await get_user(usr.hex)
+async def check_user_exists(req: Request, usr: Optional[str] = None) -> User:
+    if req.state.user:
+        user = await get_user(req.state.user.id)
+        assert user, "Logged in user has to exist."
+        g().user = user
+        return user
+
+    if not usr:
+        raise HTTPException(
+            status_code=HTTPStatus.NOT_FOUND,
+            detail="Not logged in or provided ?usr argument.",
+        )
+
+    g().user = await get_user(usr)
```

Comment:
- why is this `g().user` required, why not use a local variable?

Reply:
- the code is fragile, is a new `async` call is introduced then race conditions will occur
- by running this code:
```py
    try:
        g_user_id = g().user.id if g().user else None
        print('### before [{id, g_id]',usr, g_user_id)
    except:
        pass

    g().user = await get_user(usr) # global?
    await asyncio.sleep(1)

    g_user_id = g().user.id if g().user else None
    print('### after [{id, g_id]', usr == g_user_id, usr, g_user_id)
```

 - I have managed to get into an inconsistent state
```log
### after [{id, g_id] False bf590154da614dbc943bb1fed4cf2661 f9270eec46404a38a1fda2e8af0df34a
```

Reply:
i think we should rethink the usage of `g()` altogether, i dont see much point in it